### PR TITLE
Fix #13024, sparse & and |

### LINKED ISSUE #13024: Boolean operators don't have specialized versions for sparse matrices
Ref https://groups.google.com/d/msg/julia-users/bo6YMXzWPdA/jKEWz_rfAQAJ

The following code runs in ~15 seconds:

```jl
_column(a::AbstractSparseArray, i::Integer) = sub(a.rowval, a.colptr[i]:a.colptr[i+1]-1)

# material nonimplication
⊅(p::Bool, q::Bool) = p & !q 

function ⊅(a::SparseMatrixCSC, b::SparseMatrixCSC) 
    (m,n) = size(a) 
    resultmx = spzeros(Bool,m,n) 
    for c = 1:n 
        for r in _column(a,n) 
            # info("row $r, col $c") 
            resultmx[r,c] = ⊅(a[r,c], b[r,c]) 
        end 
    end 
    return resultmx 
end

a = sprandbool(1000000,1000000,0.0001);
b = sprandbool(1000000,1000000,0.0001);

julia> @time z = a ⊅ b;
 15.272250 seconds (1.10 M allocations: 73.870 MB)
```

Replacing `@time z = a ⊅ b` with `@time z = a & b` results in a > 10 minute execution (I stopped it after 10 minutes or so).

Interestingly enough, substituting the ⊅-equivalent `resultmx[r,c] = !b[r,c]` within the function itself also causes poor performance relative to the above code.

### PR DESCRIPTION
**Pull request by tkelman**
also fix eltype promotion in sparse max and min

### CODE REVIEW

**Comment by jiahao** [comment]
Can has $ while you're at it?

**Comment by tkelman** [comment]
Sure.

File: test/sparsedir/sparse.jl [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
+using Base.Test
+
```
Comment (yuyichao) [inline test/sparsedir/sparse.jl]: Is this just for running this as a script alone?

Comment (tkelman) [inline test/sparsedir/sparse.jl]: yeah, to make `./julia test/sparsedir/sparse.jl` work - should be harmless

**Comment by sbromberger** [comment]
Thank you both so much for this. Efficient `A | B` is critical to some of my code.

**Comment by sbromberger** [comment]
Sorry for the naive question, but what's the process for getting this into the 4.x branch? (Am I correct in assuming it won't make it into -release?)

**Comment by tkelman** [comment]
It didn't make it into RC1, but I'll probably backport it before RC2. This is slightly breaking if anyone were relying on the existing incorrect and slow output types, so my preference would be to at least let PackageEvaluator run with this in the nightlies before doing the backport. I'll write up a formal backporting policy when I get some time to do so.

**Comment by sbromberger** [comment]
thanks very much for the explanation, and I'm thrilled to hear it might make it into -release. I don't know much about how backported features are tested with PackageEvaluator but if you need an extra set of eyes/hands and don't mind the occasional silly question I'm happy to help out.

**Comment by tkelman** [comment]
whoops, added the wrong # reference to the backport commit message here. backported in #13072

**Pull request merged**